R/quotation.R

```diff
@@ -0,0 +1,363 @@
+#' Quotation
+#'
+#' @description
+#'
+#' Quotation is a mechanism by which an expression supplied as
+#' argument is captured by a function. Instead of seeing the value of
+#' the argument, the function sees the recipe (the R code) to make
+#' that value. This is possible because R [expressions][is_expr] are
+#' representable as regular objects in R:
+#'
+#' * Calls represent the action of calling a function to
+#'   compute a new value. Evaluating a call causes that value to be
+#'   computed. Calls typically involve symbols to reference R objects.
+#'
+#' * Symbols represent the name that is given to an object in a
+#'   particular context (an [environment][env]).
+#'
+#' We call objects containing calls and symbols [expressions][is_expr].
+#' There are two ways to create R expressions. First you can **build**
+#' calls and symbols from parts and pieces (see [sym()], [syms()] and
+#' [call2()]). The other way is by *quotation*, i.e. by intercepting
+#' an expression instead of evaluating it.
+#'
+#'
+#' @section User expressions versus your expressions:
+#'
+#' There are two points of view when it comes to capturing an
+#' expression:
+#'
+#' * You can capture the expressions supplied by _the user_ of your
+#'   function. This is the purpose of `ensym()`, `enexpr()` and
+#'   `enquo()` and their plural variants. These functions take an
+#'   argument name and capture the expression that was supplied to
+#'   that argument.
+#'
+#' * You can capture the expressions that _you_ supply. To this end
+#'   use `expr()`, `quo()` and their plural variants.
+#'
+#'
+#' @section Capture raw expressions:
+#'
+#' * `enexpr()` and `expr()` capture a single raw expression.
+#'
+#' * `enexprs()` and `exprs()` capture a list of raw expressions
+#'   including expressions contained in `...`.
+#'
+#' * `ensym()` and `ensyms()` are variants of `enexpr()` and
+#'   `enexprs()` that check the captured expression is either a string
+#'   (which they convert to symbol) or a symbol. If a call or data
+#'   (e.g. a number like `1L`) is supplied they throw an error.
+#'
+#' In terms of base functions, `enexpr()` corresponds to
+#' [base::substitute()] (though that function has complex semantics)
+#' and `expr()` is like [quote()] and [bquote()]. The plural variant
+#' `exprs()` is equivalent to [base::alist()].
+#'
+#'
+#' @section Capture expressions in quosures:
+#'
+#' `quo()` and `enquo()` are similar to their `expr` counterparts but
+#' box the captured expression in a wrapper called a quosure. This
```

Comment:
but capture both the expression and its environment in an object called a quosure.